File: abstract_factory/AbstractFactory.py
from abc import ABC, abstractmethod

# Abstract products
class Button(ABC):
    @abstractmethod
    def render(self):
        pass

class Checkbox(ABC):
    @abstractmethod
    def render(self):
        pass

# Concrete products
class WindowsButton(Button):
    def render(self):
        print("Rendering Windows Button")

class WindowsCheckbox(Checkbox):
    def render(self):
        print("Rendering Windows Checkbox")

class MacOSButton(Button):
    def render(self):
        print("Rendering MacOS Button")

class MacOSCheckbox(Checkbox):
    def render(self):
        print("Rendering MacOS Checkbox")

# Abstract factory
class GuiFactory(ABC):
    @abstractmethod
    def create_button(self):
        pass

    @abstractmethod
    def create_checkbox(self):
        pass

# Concrete factories
class WindowsGuiFactory(GuiFactory):
    def create_button(self):
        return WindowsButton()

    def create_checkbox(self):
        return WindowsCheckbox()

class MacOSGuiFactory(GuiFactory):
    def create_button(self):
        return MacOSButton()

    def create_checkbox(self):
        return MacOSCheckbox()

# Example application
class Application:
    def __init__(self, factory: GuiFactory):
        self.button = factory.create_button()
        self.checkbox = factory.create_checkbox()

    def render(self):
        self.button.render()
        self.checkbox.render()

# Main entry point
if __name__ == "__main__":
    os = "Windows"  # Change to "MacOS" for MacOSGuiFactory
    factory = WindowsGuiFactory() if os == "Windows" else MacOSGuiFactory()

    app = Application(factory)
    app.render()

File: creational_design_patterns/abstract_factory/AbstractFactory.java
package abstract_factory;

// Abstract products
interface Button {
    void render();
}

interface Checkbox {
    void render();
}

// Concrete products
class WindowsButton implements Button {
    @Override
    public void render() {
        System.out.println("Rendering Windows Button");
    }
}

class WindowsCheckbox implements Checkbox {
    @Override
    public void render() {
        System.out.println("Rendering Windows Checkbox");
    }
}

class MacOSButton implements Button {
    @Override
    public void render() {
        System.out.println("Rendering MacOS Button");
    }
}

class MacOSCheckbox implements Checkbox {
    @Override
    public void render() {
        System.out.println("Rendering MacOS Checkbox");
    }
}

// Abstract factory
abstract class GuiFactory {
    abstract Button createButton();

    abstract Checkbox creatCheckbox();
}

// Concrete factories
class WindowsGuiFactory extends GuiFactory {
    @Override
    Button createButton() {
        return new WindowsButton();
    }

    @Override
    Checkbox creatCheckbox() {
        return new WindowsCheckbox();
    }
}

class MacOSGuiFactory extends GuiFactory {
    @Override
    Button createButton() {
        return new MacOSButton();
    }

    @Override
    Checkbox creatCheckbox() {
        return new MacOSCheckbox();
    }
}

// Example application
class Application {
    private Button button;
    private Checkbox checkbox;

    public Application(GuiFactory factory) {
        button = factory.createButton();
        checkbox = factory.creatCheckbox();
    }

    public void render() {
        button.render();
        checkbox.render();
    }
}

public class AbstractFactory {
    public static void main(String[] args) {
        GuiFactory factory;

        String os = "Windows";
        if ("Windows".equals(os)) {
            factory = new WindowsGuiFactory();
        } else {
            factory = new MacOSGuiFactory();
        }

        Application app = new Application(factory);
        app.render();
    }
}
